From: Michael Chan <michael.chan@broadcom.com>
Date: Sat, 16 Dec 2017 03:09:41 -0500
Subject: net: Disable GRO_HW when generic XDP is installed on a device.
Patch-mainline: v4.16-rc1
Git-commit: 56f5aa77cdad1076bea0ae8ddeb74ba68ddc9502
References: bsc#1050242 FATE#322914

Hardware should not aggregate any packets when generic XDP is installed.

Cc: Ariel Elior <Ariel.Elior@cavium.com>
Cc: everest-linux-l2@cavium.com
Signed-off-by: Michael Chan <michael.chan@broadcom.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Thomas Bogendoerfer <tbogendoerfer@suse.de>
---
 net/core/dev.c |   18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -1540,6 +1540,23 @@ void dev_disable_lro(struct net_device *
 }
 EXPORT_SYMBOL(dev_disable_lro);
 
+/**
+ *	dev_disable_gro_hw - disable HW Generic Receive Offload on a device
+ *	@dev: device
+ *
+ *	Disable HW Generic Receive Offload (GRO_HW) on a net device.  Must be
+ *	called under RTNL.  This is needed if Generic XDP is installed on
+ *	the device.
+ */
+static void dev_disable_gro_hw(struct net_device *dev)
+{
+	dev->wanted_features &= ~NETIF_F_GRO_HW;
+	netdev_update_features(dev);
+
+	if (unlikely(dev->features & NETIF_F_GRO_HW))
+		netdev_WARN(dev, "failed to disable GRO_HW!\n");
+}
+
 static int call_netdevice_notifier(struct notifier_block *nb, unsigned long val,
 				   struct net_device *dev)
 {
@@ -4357,6 +4374,7 @@ static int generic_xdp_install(struct ne
 		} else if (new && !old) {
 			static_key_slow_inc(&generic_xdp_needed);
 			dev_disable_lro(dev);
+			dev_disable_gro_hw(dev);
 		}
 		break;
 
